### PR TITLE
filters: add onStreamComplete

### LINKED ISSUE #13174: log: allow extensions enrich the logs (regression)
We have some use-cases where a filter needs to set dynamic metadata when the stream ends, but before log() happens, to enrich the logs with data gathered through the lifecycle of the request (and regardless if it ended or reset).

We used to do it using onDestroy(), but now that log() happens before onDestroy, this approach doesn't work anymore (as of https://github.com/envoyproxy/envoy/pull/12850).

solution suggestion: adding another callback to `StreamFilterBase` named `onStreamCompleted` that will be called just before `log()` happens here: https://github.com/envoyproxy/envoy/blob/84f538f1ddcb5c16d98627df8b5e8c38f0f26720/source/common/http/conn_manager_impl.cc#L207-L208

This will allow extensions to enrich the logs, without the need to reason on all the cases where the stream might end (i.e. if it was reset or not).

If there are no objections, we can contribute this change.

cc @kyessenov

### PR DESCRIPTION
**Pull request by yuval-k**
Add onStreamComplete filter callback that's called before log handlers

Commit Message:
Add onStreamComplete callback to allow filters modify state before access logs are created.
Additional Description:
Risk Level: Low; added new optional callback.
Testing: Added unit-tests.
Docs Changes: N/A
Release Notes: N/A
Fixes #13174

### CODE REVIEW

File: include/envoy/http/filter.h [hunk 1]
```diff
@@ -528,12 +528,27 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
 };
 
 /**
- * Common base class for both decoder and encoder filters.
+ * Common base class for both decoder and encoder filters. Functions here are related to the
+ * lifecycle of a filter. Currently the life cycle is as follows:
+ * - All filters receive onPreDestroy()
+ * - All long handlers receive log()
```
Comment (kyessenov) [inline include/envoy/http/filter.h:534]: _log_

Comment (yuval-k) [reply include/envoy/http/filter.h]: fixed, thanks!

**Comment by yuval-k** [comment]
/retest
On the chance this is a flake.

I'm seeing gcc failed on the test I added (`'//test/common/http:conn_manager_impl_test`).
But I only see
```
2020-09-22T18:00:23.9177306Z gcc: fatal error: Killed signal terminated program cc1plus
```
And i'm not seeing a real compilation error... Any idea what can be the cause here?
full log: https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/51877/logs/108

**Comment by yuval-k** [comment]
i think the gcc and asan builds failed due to the test being too large - when I split the exact same test to a different file, the asan works (the gcc fails due to no space left on device flake).

@lizan maybe you have more context/advice on how to proceed??

**Comment by mattklein123** [comment]
CI is broken right now. You can follow along in #envoy-ci on slack.

/wait

File: include/envoy/http/filter.h [hunk 2]
```diff
@@ -528,12 +528,27 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
...
 /**
- * Common base class for both decoder and encoder filters.
+ * Common base class for both decoder and encoder filters. Functions here are related to the
+ * lifecycle of a filter. Currently the life cycle is as follows:
+ * - All filters receive onPreDestroy()
+ * - All log handlers receive log()
+ * - All filters receive onDestroy()
+ *
+ * This means:
+ * - onPreDestroy can be used to make state changes that are indented to appear in the access logs
```
Comment (kyessenov) [inline include/envoy/http/filter.h:538]: _intended_

**Comment by kyessenov** [comment]
LGTM but I don't like the test split. I think we should split it, but based on some criteria (not test_1 and test_2, since it's not clear which one stand for what).

**Comment by yuval-k** [comment]
@kyessenov I agree about the test split - the test split waf just to test if its a CI issue or not; once CI is fixed I will revert it (that's why i made the PR a draft, so this split doesn't get merged)

**Comment by lizan** [comment]
I don't like the name of this method, shall we call it onPreLog, or onStreamComplete?

**Comment by yuval-k** [comment]
@mattklein123 suggested `onPreDestroy` (https://github.com/envoyproxy/envoy/issues/13174#issuecomment-694996057), I'm fine with changing this - want to make sure we all agree on the new name.

**Comment by mattklein123** [comment]
Don't we already have `onComplete()` or similar? Now that I think about it isn't that enough?

**Comment by yuval-k** [comment]
I did see `completeRequest()` [here](https://github.com/envoyproxy/envoy/blob/42c444367f87abae02678057123d4e1863fd7f22/source/common/http/conn_manager_impl.cc#L206), but that's implemented on the ActiveStream, interacts with the StreamInfo, and cannot be overriden by the filter...

I can rename `onPreDestroy` to `onStreamComplete` if you think that's a better name?

**Comment by mattklein123** [comment]
https://github.com/envoyproxy/envoy/blob/41e65e25ed60e6d036de7f5ff610c43cfc45cbf6/include/envoy/http/filter.h#L602

**Comment by yuval-k** [comment]
the problem with this function is that it is not called when the stream ends, but when decoding ends... so it's not a good fit for this use case (i.e. it may-be be called before encodeHeaders(..) is called)

**Comment by yuval-k** [comment]
a similar problem with `encodeComplete` complete - I'm not sure if it is called if for example the client reset the connection, and no encoding is performed..

**Comment by mattklein123** [comment]
OK makes sense. I'm fine with a different name. `onStreamComplete()` seems fine. Please just document clearly when various things are called in the header file.

**Comment by yuval-k** [comment]
I added a comment to `StreamFilterBase` explaining the life-cycle, and a comment to the function itself. Please let me know if you think I missed something

File: source/common/http/conn_manager_impl.cc [hunk 3]
```diff
@@ -204,6 +204,7 @@ void ConnectionManagerImpl::doDeferredStreamDestroy(ActiveStream& stream) {
   stream.filter_manager_.disarmRequestTimeout();
 
   stream.completeRequest();
+  stream.filter_manager_.preDestroyFilters();
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc:207]: Can you be consistent here with naming?

File: include/envoy/http/filter.h [hunk 4]
```diff
@@ -528,12 +528,27 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
...
+ * Common base class for both decoder and encoder filters. Functions here are related to the
+ * lifecycle of a filter. Currently the life cycle is as follows:
+ * - All filters receive onStreamComplete()
+ * - All log handlers receive log()
+ * - All filters receive onDestroy()
+ *
+ * This means:
+ * - onStreamComplete can be used to make state changes that are intended to appear in the access
+ * logs (like streamInfo().dynamicMetadata() or streamInfo().filterState()).
+ * - onDestroy is used to cleanup all pending filter resources like pending http requests.
```
Comment (mattklein123) [inline include/envoy/http/filter.h:540]:
```suggestion
 * - onDestroy is used to cleanup all pending filter resources like pending http requests and timers.
```

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks lgtm with small comments.

/wait

File: source/common/http/conn_manager_impl.cc (hunk 3 above)
Comment (yuval-k) [reply source/common/http/conn_manager_impl.cc]: sorry about that, missed it in the search and replace

**Comment by mattklein123** [comment]
Please update the PR title and description.

/wait-any

**Comment by yuval-k** [comment]
/retest
test/extensions/filters/http/oauth2:oauth_integration_test failed in coverage - most likely a flake as i didn't touch that code..

updated pr title & description

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**